### ISSUE TITLE
reqmgr2 taskchain with different numEvents per task

### ISSUE THREAD

**Issue by anorkus** [dmwm/WMCore#7653]
HI,

@vlimant 
we have specific case where a requests starting from same GS, but doing DR with multiple PU scenarios. In this specific case GS+3 DR's:
https://cmsweb.cern.ch/reqmgr2/data/request?name=pdmvserv_task_EXO-PhaseIIFall16GS82-00046__v1_T_170216_201031_2250

The number_of_events will be 20k as defined in GS request, and all further requests will run over same number of events. Can we specify (lets say) 5k events for Task{2;3} and these DR Tasks would run only on subset of GS events?

In this case we wouldn't need to submit a task with GS+single DR. And then when workflow completes, submit multiple DR's on subset of GS.

Best regards,
    Antanas Norkus

**Comment by amaltaro** [comment]
You cannot run on only part of the output of the previous task. Number of events is also pointless when a task has an input dataset, since it uses the whole dataset (unless you specify blocks/run/lumilist, which can only be used for the first task in a request).

Since you don't have any input dataset on this case, you should just replace
"RequestNumEvents": 20000
by
"RequestNumEvents": 5000

and then the 3 DR tasks would run on 5k events only :-)

**Comment by anorkus** [comment]
yeah, but GS would also generate only 5000 events too ;) which we do not want.

**Comment by vlimant** [comment]
@amaltaro I wasn't sure RequestNumEvents would work for internal tasks and suggested the GH.
if RequestNumEvents is supported for limiting the number of events processed in a specific "branch" of the chain, then perfect

**Comment by amaltaro** [comment]
Antanas, then you need to request 20k events and get the 20k events in the DR output. There is no other option.

Jean-Roch, `RequestNumEvents` is only meaningful for the top level task AND if it's a production task (no input data).

**Comment by vlimant** [comment]
@amaltaro the GH issue is for making the RequestNumEvents functioning to do sub-statistics within the chain : i.e. create only a limit number of jobs to bit as close as >RequestNumEvents.
I think it makes sense

**Comment by amaltaro** [comment]
This would be a major WMAgent development. The "operational" motivation is reasonable, now why you would need to have much more events in GS than in DR is not clear yet.

BTW, we need to evaluate the cost-benefit ratio. If this is going to be rarely used, I'd rather invest time on things that have a bigger impact for CMS.

**Comment by vlimant** [comment]
In the example provided it's GS->DR1+DR2+DR3, DR1 and DR2 have full stats, but DR3 needs smaller stats.
In a view to using chains more and more, we should support this.
I say, in medium priority, that we have to estimate how deep we have to go to have this happen.

**Comment by anorkus** [comment]
well, with more Upgrade high PU requests coming, we produce GS+DR_noPU for whole GS generated events. But then if we want to do DR with 140PU or 200PU, we want to run less events due to high PU. This means we end up in need of this feature: where subset tasks run on sub-statistics.
or 
we submit GS+DR_noPU, wait for workflow to complete/announce, then submit the high PU DR's separately. This brings more manual work (especially when we have issues with rerunning/extensions, invalidations etc.) also brings delay to analysis when high PU datasets would be available.

Sure, we can run high PU for full dataset, but then we would not use production system efficiently (over producing events).

**Comment by hufnagel** [comment]
If DR3 needs smaller stats put a filter in the DR3 configuration. Doing this in the agent goes against basic design principles of the agent.

**Comment by hufnagel** [comment]
That being said, @amaltaro , there is an "easy" way to do this on the agent side. We would need a custom job splitter for Task3 (we can customize job splitting per task) that configures jobs with input filters, processing only some fraction of an input file or only selected input files. It's still somewhat hacky and IMO a filter on input would be better (also to give a better statistical sample), but it should work.

**Comment by vlimant** [comment]
@hufnagel adding the filter on top is rather intrusive of the configuration. if there is a way to have to agent throttle the amount of files processed based on the recorded number of events, overall it would be less intrusive and makes more sense. but I appreciate the cost benefit

**Comment by hufnagel** [comment]
The only way I can see this being done is in the job splitter and the job splitter doesn't have a global view of the task. It only sees a random subset of the DR3 task input. (whatever is available to schedule new jobs on when it runs). Within that random subset it can apply filters (ie. have 3 files, lets only process 1). If the subset is too small, the selection breaks down though. What do you do if you only have one file ? So doing this in the job splitter is going to be very imprecise, which might be ok if the goal here is to save wasting resources and you don't care too much if you end up with 25%, 30% or 35% of the events processed.

Whether a file based selected or a lumi within file based selection works best I can't predict based of the top of my hat. Depends how the DR3 task input data is organized (do you expect multiple lumis per file ? do you expect multiple files to be available on average in a job splitting run ?). I don't know enough about the workflows to say.

If you are way off and implement a file or lumi based filter in the job splitting and then on average the job splitter only ever sees a single file with a single lumi you would still process almost all the data.

If you don't care about screwing up the accounting an event based filter should always work, but that will screw up the lumi accounting. The DR3 output will claim to contain lumis that are not complete. Which I think would be bad if you want to analyze this data and make sense of it.

**Comment by hufnagel** [comment]
Apart from the technicalities, I still think doing this in the agent opens a can of worms. You would expose a setting (fraction of files processed or fraction of lumis processed in task) that

a) is yet another switch by which workflow requesters can shot their legs off (if they want everything)
b) which is highly dependent on runtime behavior of workflow (number of new files produced per site within the JobCreator polling cycle, number of lumis in files) and where you have no idea what you are actually going to get upfront

IMO having a custom configuration for the DR3 task with an event filter (only process every Nth) event is the only reliable way to get this done and it can be done now without any agent changes.

**Comment by ericvaandering** [comment]
I think this falls into possible use cases for the "do things by work" refactoring. One thing I envisage is being able to have a separate lumi mask for each step and/or task. So in this case, you'd have a parent task which generated everything and a mask on the DR3 step that could select only a portion of the output as input. But having this is many months away.
